pkg/scheduler/plugins/gang/gang.go

```diff
@@ -0,0 +1,437 @@
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+var (
+	timeNowFn = time.Now
+)
+
+const (
+	GangFromCrd        string = "GangFromCrd"
+	GangFromAnnotation string = "GangFromAnnotation"
+)
+
+// Gang  basic gang info recorded in gangCache:
+type Gang struct {
+	Name       string
+	WaitTime   time.Duration
+	CreateTime time.Time
+	//strict-mode or non-strict-mode
+	Mode              string
+	MinRequiredNumber int
+	TotalChildrenNum  int
+	GangGroup         []string
+	Children          map[string]*v1.Pod
+	//pods that have already assumed(waiting in Permit stage)
+	WaitingForBindChildren map[string]*v1.Pod
+	//pods that have already bound
+	BoundChildren map[string]*v1.Pod
+	//if assumed  pods number has reached to MinRequiredNumber
+	ResourceSatisfied bool
+
+	//if the gang should be passed at PreFilter stage(Strict-Mode)
+	ScheduleCycleValid bool
+	//these fields used to count the cycle
+	ScheduleCycle            int
+	ChildrenScheduleRoundMap map[string]int
+
+	GangFrom    string
+	HasGangInit bool
+
+	lock sync.Mutex
+}
+
+func NewGang(gangName string) *Gang {
+	return &Gang{
+		Name:                     gangName,
+		CreateTime:               timeNowFn(),
+		WaitTime:                 0,
+		Mode:                     extension.GangModeStrict,
+		Children:                 make(map[string]*v1.Pod),
+		WaitingForBindChildren:   make(map[string]*v1.Pod),
+		BoundChildren:            make(map[string]*v1.Pod),
+		ScheduleCycleValid:       true,
+		ScheduleCycle:            1,
+		ChildrenScheduleRoundMap: make(map[string]int),
+		GangFrom:                 GangFromAnnotation,
+		HasGangInit:              false,
+	}
+}
+
+/******************Pod\PodGroup Event BEG******************/
+func (gang *Gang) TryInitByPodConfig(pod *v1.Pod, args *schedulingconfig.GangSchedulingArgs) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	if gang.HasGangInit {
+		return
+	}
+
+	minRequiredNumber, err := strconv.Atoi(pod.Annotations[extension.AnnotationGangMinNum])
+	if err != nil {
+		klog.Errorf("pod's annotation MinRequiredNumber illegal, gangName:%v, value:%v",
+			gang.Name, pod.Annotations[extension.AnnotationGangMinNum])
+		return
+	}
+	gang.MinRequiredNumber = minRequiredNumber
+
+	totalChildrenNum, err := strconv.Atoi(pod.Annotations[extension.AnnotationGangTotalNum])
+	if err != nil {
+		klog.Errorf("pod's annotation TotalNumber illegal, gangName:%v, value:%v",
+			gang.Name, pod.Annotations[extension.AnnotationGangTotalNum])
+		totalChildrenNum = minRequiredNumber
+	} else if totalChildrenNum != 0 && totalChildrenNum < minRequiredNumber {
+		klog.Errorf("pod's annotation TotalNumber cannot less than minRequiredNumber, gangName:%v, TotalNumber:%v,minRequiredNumber:%v",
+			gang.Name, pod.Annotations[extension.AnnotationGangTotalNum], minRequiredNumber)
+		totalChildrenNum = minRequiredNumber
+	}
+	gang.TotalChildrenNum = totalChildrenNum
+
```

Comment:
gang.CreateTime = pod.CreationTimestamp.Time

Reply:
done

Reply:
What I'm not sure about is whether it's possible to set the creation time on first initialization from a pod that is relatively new in time dimension. Should we use the oldest time within the same gang or the newest time? Or that it doesn't matter?

Reply:
we set gang a field called "TimeoutStartTime",which means from this time we start coutning the timeout,it depends on the first pod coming to the Permit stage and decided to Wait at this stage,we set the TimeoutStartTime of that moment.